storagenode/pieces/store.go

```diff
@@ -46,42 +46,93 @@ type Info struct {
-	SpaceUsed(ctx context.Context) (int64, error)
-	// CalculatedSpaceUsed calculates disk space used by all pieces
-	CalculatedSpaceUsed(ctx context.Context) (int64, error)
-	// SpaceUsedBySatellite calculates disk space used by all pieces by satellite
-	SpaceUsedBySatellite(ctx context.Context, satelliteID storj.NodeID) (int64, error)
-	// GetExpired gets orders that are expired and were created before some time
+	// GetExpired gets piece IDs stored with storage format V0 that expire or have expired
+	// before the given time
 	GetExpired(ctx context.Context, expiredAt time.Time, limit int64) ([]ExpiredInfo, error)
+	// ForAllV0PieceIDsOwnedBySatellite executes doForEach for each locally stored piece, stored
+	// with storage format V0 in the namespace of the given satellite. If doForEach returns a
+	// non-nil error, ForAllV0PieceIDsOwnedBySatellite will stop iterating and return the error
+	// immediately.
+	ForAllV0PieceIDsOwnedBySatellite(ctx context.Context, blobStore storage.Blobs, satellite storj.NodeID, doForEach func(StoredPieceAccess) error) error
+}
+
+// V0PieceInfoDBForTest is like V0PieceInfoDB, but adds on the Add() method so
+// that test environments with V0 piece data can be set up.
+type V0PieceInfoDBForTest interface {
+	V0PieceInfoDB
+
+	// Add inserts Info to the database. This is only a valid thing to do, now,
+	// during tests, to replicate the environment of a storage node not yet fully
+	// migrated to V1 storage.
+	Add(context.Context, *Info) error
+}
+
+// StoredPieceAccess allows inspection and manipulation of a piece during iteration with
+// ForAllPieceIDsOwnedBySatellite-type methods
+type StoredPieceAccess interface {
+	storage.StoredBlobAccess
+
+	// PieceID gives the pieceID of the piece
+	PieceID() storj.PieceID
+	// Satellite gives the nodeID of the satellite which owns the piece
+	Satellite() (storj.NodeID, error)
+	// ContentSize gives the size of the piece content (not including the piece header, if
+	// applicable)
+	ContentSize(ctx context.Context) (int64, error)
+	// CreationTime returns the piece creation time as given in the original PieceHash (which is
+	// likely not the same as the file mtime). For non-FormatV0 pieces, this requires opening
+	// the file and unmarshaling the piece header. If exact precision is not required, ModTime()
+	// may be a better solution.
+	CreationTime(ctx context.Context) (time.Time, error)
+	// ModTime returns a less-precise piece creation time than CreationTime, but is generally
+	// much faster. For non-FormatV0 pieces, this gets the piece creation time from to the
+	// filesystem instead of the piece header.
+	ModTime(ctx context.Context) (time.Time, error)
 }
 
 // Store implements storing pieces onto a blob storage implementation.
 type Store struct {
-	log   *zap.Logger
-	blobs storage.Blobs
+	log            *zap.Logger
+	blobs          storage.Blobs
+	v0PieceInfo    V0PieceInfoDB
+	expirationInfo PieceExpirationDB
+
+	reservedSpace int64
```

Comment:
can we add a comment that this is just used for testing?

Reply:
Yes. I do have this (and a few other clarifications you asked for in the walkthrough) in some pending changes. I will get that separated nicely into commits and push it.